tests/test_utils/test_misc_tools.py

```diff
@@ -0,0 +1,87 @@
+# Copyright 2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytest
+from copy import deepcopy
+from tests.mock_classes.general_mocks import *
+from mrmustard.utils.misc_tools import *
+
+
+###############################################################################
+##########################   duck_type_checker   ##############################
+@pytest.mark.parametrize("cls_instance", [MockAnimal(), MockFruit()])
+def test_dtc_returns_true_when_given_same_object(cls_instance):
+    other_instance = deepcopy(cls_instance)
+    assert duck_type_checker(cls_instance, other_instance) == True
+
+
+@pytest.mark.parametrize("obj_a, obj_b", [(MockAnimal(), MockAnimal()), (MockFruit(), MockFruit())])
+def test_dtc_returns_true_when_given_different_object_of_same_type(obj_a, obj_b):
+    assert duck_type_checker(obj_a, obj_b)
+
+
+@pytest.mark.parametrize(
+    "obj_type_a, obj_type_b", [(MockAnimal(), MockFruit()), (MockFruit(), MockAnimal())]
+)
+def test_dtc_returns_same_bool_irrelevant_of_object_order(obj_type_a, obj_type_b):
+    assert duck_type_checker(obj_type_a, obj_type_b) == duck_type_checker(obj_type_b, obj_type_a)
+
+
+@pytest.mark.parametrize(
+    "obj_type_a, obj_type_b, truth_val",
+    [
+        (MockAnimal(), MockFruit(), False),
+        (MockAnimal(), MockAnimal(), True),
+        (MockFruit(), MockFruit(), True),
+        (MockFruit(), MockAnimal(), False),
+    ],
+)
+def test_dtc_returns_correct_bool_with_objects_of_different_or_same_type(
+    obj_type_a, obj_type_b, truth_val
+):
+    assert duck_type_checker(obj_type_a, obj_type_b) == truth_val
+
+
+@pytest.mark.parametrize("x, y", [(1, 1.0), (1.0, "s"), ("s", 2.0), (True, 1.0), (MockAnimal(), 1)])
+def test_dtc_raises_TypeError_when_given_objects_without_dicts(x, y):
+    with pytest.raises(TypeError):
+        duck_type_checker(x, y)
+
+
+##########################   general_factory   ##############################
+@pytest.mark.parametrize(
+    "cls, args, kwargs",
+    [
+        (MockAnimal, (1954, False), {"colour": "red"}),
+        (MockFruit, (), {}),
+        (MockNoDefaultParams, (), {"a": 1954, "b": "no"}),
+        (MockNoDefaultParams, (1954, "yes"), {}),
+    ],
+)
+def test_gfactory_returns_instance_of_correct_class(cls, args, kwargs):
+    assert isinstance(general_factory(cls, *args, **kwargs), cls) == True
```

Comment:
I still wonder what prevents you from doing `cls(*args, **kwargs)` directly 🤔 The behaviour expected from this and the following test would be exactly the same (i.e. create an instance and reaise an error when extra/wrong/unexpected args are given)

Reply:
What prevents me from doing it? Uhm, I'd say nothing beyond personal tastes xD. I can do that if you'd rather. Since that's exactly the one and only line of code contained in the `general_factory` anyway :').
Now my rationale was to still implement this general factory for 2 reasons:
1) There are no factory methods currently in MM afaik, and as we had briefly discussed it would be great to introduce them, so my thoughts were that this would be a first baby step in introducing the general concept of a factory there.
2) When we add more factories, we can probably inherit from the general_factory tests by replacing this method by XXfactory, which we couldn't do if I was using just the barebone line of code. I agree this would have made more sense with more than 2 tests ^^'.

So those were my `why`. They're not necessarily good/accurate ^^'.

So let me know your thoughts, I can remove the factory and the tests if you'd like, no problem.

Reply:
Perhaps before deciding if we should keep it or remove it it'd be good to look at its usage pattern. Can you point me to some code that uses this factory?

Reply:
For now I only use it in the tests.
if you go to the stage1_MVP_data branch, you'll find it in the fixture called DATA in all files, you can also find it inside some of the tests when I need to build a specific instance of the desired type on the fly, eg: in `test_matvec_data.py` the test function `if_coeffs_not_given_they_are_equal_to_1` calls it. There are a few other instances such as this latter one.